Fix native toplevel on Windows

dra27 opened the pull request:
Setting `OCAML_FLEXLINK` allows `make natruntop` to work with the four native Windows ports.

dra27 committed:
Detect rlwrap in configure

Cygwin's rlwrap doesn't work with native console executables which
breaks the runtop target if rlwrap is installed. Move the detection of
rlwrap to configure and add an explicit test to disable it for the
native Windows ports.

gasche reviewed (commented):
The way FLEXLINK_ENV is used for other commands is in Makefile.common, by defining a variable FOO_CMD which wraps "everything needed to invoke FOO as a command".

Wouldn't it make sense to do the same here, by defining RUNTOP_CMD and NATRUNTOP_CMD in Makefile.common?

We could have something like

```make
RUNTOP:=$(EXTRAPATH) $(ROOTDIR)/runtime/ocamlrun$(EXE) ./ocaml$(EXE)
NATRUNTOP:=$(EXTRAPATH) $(ROOTDIR)/ocamlnat$(EXE)
RUNTOPFLAGS= -nostdlib -I stdlib -I toplevel -noinit $(TOPFLAGS) $(TOPINCLUDES)

...
RUNTOP_CMD:=$(FLEXLINK_ENV) $(RUNTOP)
NATRUNTOP_CMD:=$(FLEXLINK_ENV) $(NATRUNTOP)
```

and then at the callsites, $(NATRUNTOP_CMD) $(RUNTOPFLAGS).

dra27 commented:
In principle I agree, but that only reason we define `NATRUNTOP` at all is because it's used twice in the single command which needs it in because we don't have proper detection for rlwrap in configure (which is where that test really belongs). If `rlwrap` were detected "properly", I wouldn't expect to use a `NATRUNTOP` variable at all (the command for `natruntop` would simply be in that recipe.

gasche commented:
I don't understand the relation to `rlwrap`. I think you are suggesting that my refactoring wouldn't work well with rlwrap, but I'm not sure why. (I wondered if the resulting `rlwrap $(FLEXLINK_ENV) $(EXTRAPATH) ...` would be the wrong order, but your PR also seems to pass at least the FLEXLINK_ENV part after rlwrap in the same way.)

In any case, my meta point is that it would be nice if the handling of FLEXLINK_ENV stuff remained in Makefile.common only. Can you propose an approach that works and preserves this form of locality?

dra27 commented:
Hah - while checking a response, I find that there’s a problem with what I propose (and with what you propose!) ... then I discover that running the Windows native ocamlnat with rlwrap is a disaster (which needs to be avoided, therefore!), so I shall return to this properly anon.!

dra27 committed:
Fix natruntop with bootstrapped flexlink

dra27 committed:
Fix caml_globalsym on Windows

A bug in flexdll means that caml_globalsym didn't work correctly on
native Windows. The bug is also fixed in flexdll, but it's also possible
to work-around it in OCaml.

dra27 commented:
@gasche - it's indeed correct that `FLEXDLL_ENV` like `EXTRAPATH` must be passed before rlwrap is called. I was testing on a system without rlwrap so hadn't spotted this.

I have now:
- Moved the detection of rlwrap to `configure` so now `RLWRAP` is `rlwrap` if the command is both available **and appropriate**. It turns out rlwrap does not interact at all with native Windows executables (it's also unnecessary - `cmd` has a certain level of default readline-like support and one can also install Clink in order to have complete readline support by default).
- `FLEXLINK_ENV` is now passed in the correct place with `EXTRAPATH`
- Having run this, I found a long-standing bug in native dynlink for Windows. As far as I can tell, ocamlnat has been broken for native Windows since 4.06 by the bug fixed in alainfrisch/flexdll#90. There is a trivial fix which can be applied in win32.c which I've done here. For native Windows, `flexdll_dlopen` is a stub which converts its argument to Unicode and calls `flexdll_wdlopen`. The bug in flexdll is that it wasn't handling `NULL` correctly - the work-around fix in OCaml is just to call `flexdll_wdlopen` directly.

dra27 commented:
Rebased - which means that the submodule will now pick up the switch to flexdll 0.39 which includes the fix mentioned above.

Is this otherwise good to go?

Octachron reviewed (approved):
The Windows fix looks good to me, and the detection of rlwrap at configuration time seems fine.

dra27 commented:
Thanks, @Octachron! Rebased to fix the conflict

gasche commented:
@Octachron on October 7 we milestoned (?) this for 4.12, do you still plan to include in the release? (That sounds reasonable to me.)

Octachron commented:
Yes, I still plan to include it since this is a small fix for an experimental tool.
